### PR TITLE
Add update product actions to product data store

### LINKED ISSUE #33076: Add/edit simple products via products data store API (experimental)
### Description

The ability to add/edit a simple product through the products data store.

Key properties:
- Name
- Description
- Slug
- Main image
- Price
- TBD: others?

This work will be extended over time to eventually support any type of product.

For now, this should be marked as experimental/internal as we iterate on the API so that it is clear that this incomplete and subject to change.

### Acceptance criteria

- Data store updates
- Unit tests
- Documentation with examples

### References

- REST API endpoint: https://woocommerce.github.io/woocommerce-rest-api-docs/#create-a-product

### PR DESCRIPTION
**Pull request by joshuatf**
### All Submissions:

-   [x] Have you followed the [WooCommerce Contributing guideline](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md)?
-   [x] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/)?
-   [x] Have you checked to ensure there aren't other open [Pull Requests](../../pulls) for the same update/change?

### Changes proposed in this Pull Request:

Closes #33076  .

### How to test the changes in this Pull Request:

1. Go to `wp-admin/edit.php?post_type=product`
2. Pick an ID of a recently published product
3. Open your browser’s console
4. Run the following.  You may need to run again after the fetch has been completed.
```
5. wp.data.select('wc/admin/products').getProducts({ status: 'publish'}, undefined);
```
6. Note the product you picked.
7. Enter the following, replacing the `123` ID with your product ID:
```
wp.data.dispatch('wc/admin/products').updateProduct( 123, {
    name: ‘Holy smokes!’,
} );
```
8. Run the getProducts selector again:
```
wp.data.select('wc/admin/products').getProducts({ status: 'publish'}, undefined);
```
9. Note the product has been updated
10. Refresh your products page and note the product updates are persisted 

### Other information:

-   [x] Have you added an explanation of what your changes do and why you'd like us to include them?
-   [x] Have you written new tests for your changes, as applicable?
-   [x] Have you successfully run tests with your changes locally?
-   [x] Have you created a changelog file for each project being changed, ie `pnpm nx changelog <project>`?

### FOR PR REVIEWER ONLY:

-   [ ] I have reviewed that everything is sanitized/escaped appropriately for any SQL or XSS injection possibilities. I made sure Linting is not ignored or disabled.

### CODE REVIEW

**Comment by botwoo** [comment]
:bar_chart: Test reports for this pull request have been published and are accessible through the following links:

- [REST API Test Report](https://woocommerce.github.io/woocommerce-test-reports/pr/33282/api/)
- E2E Test Report (TBC)

Latest commit referenced in the reports: `Omit read only properties instead of picking mutable properties` d9dccc800df86d8782b748226f38ea5a17234dec
This comment will automatically be updated with the latest referenced commit when you push new changes to this pull request.

---

_Visit the [WooCommerce Test Reports](https://woocommerce.github.io/woocommerce-test-reports) homepage to view all published reports. See the [FAQs](https://github.com/woocommerce/woocommerce-test-reports/wiki/FAQs) page if you're having problems accessing them._

File: packages/js/data/src/products/actions.ts [hunk 1]
```diff
@@ -49,6 +49,23 @@ export function createProductError(
 	};
 }
 
+function updateProductSuccess( id: number, product: Partial< Product > ) {
+	const obj = {
```
Comment (louwie17) [inline packages/js/data/src/products/actions.ts]: Probably just use return here instead of setting it to an object.

**Review by louwie17 — DISMISSED** [review]
Left one minor comment but approving, let me know if you want me to re-approve after addressing the small suggestion.
Nice work and thanks for the additional tests :)

File: packages/js/data/src/products/actions.ts (hunk 1 above)
Comment (joshuatf) [reply packages/js/data/src/products/actions.ts]: Ah yes, this was left over from testing a bug.  Will update this, thanks for the catch!

**Comment by joshuatf** [comment]
@louwie17 Rebased this PR and included the `MutableProperties` ( https://github.com/woocommerce/woocommerce/pull/33282/commits/f0e2329eeeafba1a773854515a654c0d18df4a6d )created in the "add products" PR.  Could you give this another look?

**Review by louwie17 — DISMISSED** [review]
Approving this as it still tested well.
Although @joshuatf the `MutableProperties` list looks rather short -> https://github.com/woocommerce/woocommerce/blob/trunk/packages/js/data/src/products/types.ts#L43-L53
I am pretty sure we can update quite a few additional properties. Instead of having `MutableProperties` should we have `ImmutableProperties` instead.
Then you can just use `Omit<` instead of `Pick<`.
This could be done in a separate, given that code is already merged.

**Comment by joshuatf** [comment]
@louwie17 These were picked from those currently in the `Product` type.  I think we should wait to flesh those out more before adding to this list, unless the extra product properties are somewhere else I'm not seeing.

**Comment by louwie17** [comment]
> @louwie17 These were picked from those currently in the `Product` type. I think we should wait to flesh those out more before adding to this list, unless the extra product properties are somewhere else I'm not seeing.

Aaah yes, I will take the blame for that, I indeed hadn't added every single `product` attribute to the Product type yet, aside from the items it inherits from `Schema.Post`.
In the long term I do think it would be easier to keep track of the `readonly`/immutable fields versus the mutable fields, given the latter will be much longer.

**Comment by joshuatf** [comment]
>Aaah yes, I will take the blame for that, I indeed hadn't added every single product attribute to the Product type yet, aside from the items it inherits from Schema.Post.

I think that's fair.  The original issue was only meant to include a small subset of properties and the rest can be added as we go.

>In the long term I do think it would be easier to keep track of the readonly/immutable fields versus the mutable fields, given the latter will be much longer.

You're probably right.  Updated this to `ReadOnlyProperties` in https://github.com/woocommerce/woocommerce/pull/33282/commits/d9dccc800df86d8782b748226f38ea5a17234dec.

**Review by louwie17 — APPROVED** [review]
Thanks for updating it to `ReadOnlyProperties`, looks great, nice work on this @joshuatf 🎉

**Pull request merged**